### PR TITLE
ASoC: rt1011: fix KASAN out-of-bounds bug in find_next_bit()

### PR DESCRIPTION
**Pull request by fredoh9**
find_next_bit() try to access out-of-bounds of unsigned int variables,
rx_mask and tx_mask. Replace with unsigned long variable to avoid
pointer casting. This is verified with KASAN.

Return value of find_next_bit() is unsigned long, this is irrelevant
with the KASAN bug. But when down casting happens, at least better match
with 'unsigned'.

KASAN bug fixed,
[  170.777592] ==================================================================
[  170.777603] BUG: KASAN: stack-out-of-bounds in _find_next_bit.constprop.0+0x3e/0xf0
[  170.777608] Read of size 8 at addr ffff888202d6f530 by task aplay/1757
[  170.777615] CPU: 0 PID: 1757 Comm: aplay Not tainted 5.7.0-rc7-test+ #3
[  170.777617] Hardware name: Google Helios/Helios, BIOS  01/21/2020
[  170.777618] Call Trace:
[  170.777624]  dump_stack+0x76/0xa0
[  170.777629]  print_address_description.constprop.0.cold+0xd3/0x43e
[  170.777634]  ? _raw_spin_lock_irqsave+0x7b/0xd0
[  170.777637]  ? _raw_spin_trylock_bh+0xf0/0xf0
[  170.777640]  ? _find_next_bit.constprop.0+0x3e/0xf0
[  170.777643]  __kasan_report.cold+0x37/0x86
[  170.777647]  ? regmap_readable+0xb0/0xc0
[  170.777650]  ? _find_next_bit.constprop.0+0x3e/0xf0
[  170.777652]  kasan_report+0x38/0x50
[  170.777656]  _find_next_bit.constprop.0+0x3e/0xf0
[  170.777663]  rt1011_set_tdm_slot+0x125/0x5d0 [snd_soc_rt1011]
[  170.777669]  ? rt1011_hw_params+0x2f0/0x2f0 [snd_soc_rt1011]
[  170.777672]  ? mutex_unlock+0x18/0x40
[  170.777691]  snd_soc_dai_set_tdm_slot+0x152/0x210 [snd_soc_core]
[  170.777710]  ? snd_soc_dai_set_fmt+0x80/0x80 [snd_soc_core]
[  170.777717]  cml_rt1011_hw_params+0x20c/0x320 [snd_soc_cml_rt1011_rt5682]
[  170.777735]  snd_soc_link_hw_params+0x59/0x90 [snd_soc_core]
[  170.777752]  soc_pcm_hw_params+0x42c/0xb10 [snd_soc_core]
[  170.777771]  ? soc_pcm_open+0xbd0/0xbd0 [snd_soc_core]

Signed-off-by: Fred Oh <fred.oh@linux.intel.com>

fixes: https://github.com/thesofproject/linux/issues/2176

### CODE REVIEW

File: sound/soc/codecs/rt1011.c [hunk 1]
```diff
@@ -1801,7 +1801,10 @@ static int rt1011_set_tdm_slot(struct snd_soc_dai *dai,
 	struct snd_soc_dapm_context *dapm =
 		snd_soc_component_get_dapm(component);
 	unsigned int val = 0, tdm_en = 0, rx_slotnum, tx_slotnum;
-	int ret = 0, first_bit, last_bit;
+	unsigned long rxmask = rx_mask;
+	unsigned long txmask = tx_mask;
+	int ret = 0;
```
Comment (plbossart) [inline sound/soc/codecs/rt1011.c]: move last in declarations

File: sound/soc/codecs/rt1011.c [hunk 2]
```diff
@@ -1801,7 +1801,10 @@ static int rt1011_set_tdm_slot(struct snd_soc_dai *dai,
 	struct snd_soc_dapm_context *dapm =
 		snd_soc_component_get_dapm(component);
 	unsigned int val = 0, tdm_en = 0, rx_slotnum, tx_slotnum;
-	int ret = 0, first_bit, last_bit;
+	unsigned long rxmask = rx_mask;
+	unsigned long txmask = tx_mask;
```
Comment (plbossart) [inline sound/soc/codecs/rt1011.c]: maybe rx_mask_l and tx_mask_l to be explicit?

File: sound/soc/codecs/rt1011.c [hunk 3]
```diff
@@ -1801,7 +1801,10 @@ static int rt1011_set_tdm_slot(struct snd_soc_dai *dai,
 	struct snd_soc_dapm_context *dapm =
 		snd_soc_component_get_dapm(component);
 	unsigned int val = 0, tdm_en = 0, rx_slotnum, tx_slotnum;
-	int ret = 0, first_bit, last_bit;
```
Comment (plbossart) [inline sound/soc/codecs/rt1011.c]: comment on commit message. You don't need to include the KASAN trace, you can just copy one line 
[  170.777603] BUG: KASAN: stack-out-of-bounds in _find_next_bit.constprop.0+0x3e/0xf0

also maybe simplify the text a bit, e.g. 

KASAN throws the following warning in rt1011.c:
[  170.777603] BUG: KASAN: stack-out-of-bounds in _find_next_bit.constprop.0+0x3e/0xf0

find_next_bit() relies on unsigned long pointer arguments, but this driver uses a type cast that generates the KASAN warning. Copy the tdm_slot arguments in temporary variables and avoid casting pointers to make the warning go away.

Comment (fredoh9) [reply sound/soc/codecs/rt1011.c]: Thank you for the suggestion and reviewing this. Updated with your comments.

File: sound/soc/codecs/rt1011.c [hunk 4]
```diff
@@ -1801,7 +1801,10 @@ static int rt1011_set_tdm_slot(struct snd_soc_dai *dai,
...
+	int ret = 0;
+	unsigned int first_bit, last_bit;
```
Comment (ranj063) [inline sound/soc/codecs/rt1011.c]: @fredoh9 do we need to change the type for these? I know find_nex_bit() returns an unsigned long. So if we're changing, it it should be a separate patch and unsigned long right?

Comment (fredoh9) [reply sound/soc/codecs/rt1011.c]: This is not required change. I should just drop it. Will update again.

**Comment by fredoh9** [comment]
For failure in device test in multiple-pipeline-capture.sh with CML_MANTIS_HDA, while multiple-pipeline-playback is passed. The test case was 3x arecord + 1x aplay, one of instance is not able to run, but there is no log to dig more. I don't think the failure is related with with this PR.

`2020-06-09 22:05:36 UTC [REMOTE_ERROR] Target pipeline count: 4, current process count: 3`

File: sound/soc/codecs/rt1011.c [hunk 5]
```diff
@@ -1848,8 +1850,8 @@ static int rt1011_set_tdm_slot(struct snd_soc_dai *dai,
 	}
 
 	/* Rx slot configuration */
-	rx_slotnum = hweight_long(rx_mask);
-	first_bit = find_next_bit((unsigned long *)&rx_mask, 32, 0);
+	rx_slotnum = hweight_long(rx_mask_l);
+	first_bit = find_next_bit(&rx_mask_l, 32, 0);
```
Comment (lyakh) [inline sound/soc/codecs/rt1011.c]: `find_next_bit()` is really for finding set bits in an array of longs. Here I'd just use `__ffs(rx_mask_l)`

File: sound/soc/codecs/rt1011.c [hunk 6]
```diff
@@ -1891,9 +1893,9 @@ static int rt1011_set_tdm_slot(struct snd_soc_dai *dai,
 	}
 
 	/* Tx slot configuration */
-	tx_slotnum = hweight_long(tx_mask);
-	first_bit = find_next_bit((unsigned long *)&tx_mask, 32, 0);
-	last_bit = find_last_bit((unsigned long *)&tx_mask, 32);
+	tx_slotnum = hweight_long(tx_mask_l);
+	first_bit = find_next_bit(&tx_mask_l, 32, 0);
```
Comment (lyakh) [inline sound/soc/codecs/rt1011.c]: ditto

File: sound/soc/codecs/rt1011.c [hunk 7]
```diff
@@ -1891,9 +1893,9 @@ static int rt1011_set_tdm_slot(struct snd_soc_dai *dai,
...
+	last_bit = find_last_bit(&tx_mask_l, 32);
```
Comment (lyakh) [inline sound/soc/codecs/rt1011.c]: `__fls()` but also see if any of the masks can be 0 and what you should do then.

File: sound/soc/codecs/rt1011.c (hunk 5 above)
Comment (fredoh9) [reply sound/soc/codecs/rt1011.c]: @lyakh thank you for reviewing this. __ffs(), __fls() seems fancy and fast function!
I printed the return value to compare with current functions.

I found two mismatch.
DEBUG:TX 4 find_next_bit=0 __ffs=2
DEBUG:TX 8 find_next_bit=1 __ffs=3
Test results are all good though. Looking more closely.
```
[   81.412639] DEBUG:RX 1 find_next_bit=0 __ffs=0
[   81.412816] DEBUG:TX 4 find_next_bit=0 __ffs=2
[   81.412816] DEBUG:TX 4 find_last_bit=2 __fls=2
[   81.414490] DEBUG:RX 2 find_next_bit=1 __ffs=1
[   81.414844] DEBUG:TX 8 find_next_bit=1 __ffs=3
[   81.414845] DEBUG:TX 8 find_last_bit=3 __fls=3
[   81.416656] DEBUG:RX 1 find_next_bit=0 __ffs=0
[   81.416850] DEBUG:TX 1 find_next_bit=0 __ffs=0
[   81.416851] DEBUG:TX 1 find_last_bit=0 __fls=0
[   81.418633] DEBUG:RX 2 find_next_bit=1 __ffs=1
[   81.418990] DEBUG:TX 2 find_next_bit=1 __ffs=1
[   81.418992] DEBUG:TX 2 find_last_bit=1 __fls=1
```

Comment (fredoh9) [reply sound/soc/codecs/rt1011.c]: Oops, debug printing has an issue, retesting.

Comment (fredoh9) [reply sound/soc/codecs/rt1011.c]: My bad, debug print had an issue. Looks good, will push the update.

```
[   64.678811] DEBUG:RX 1 find_next_bit=0 __ffs=0
[   64.678986] DEBUG:TX 4 find_next_bit=2 __ffs=2
[   64.678987] DEBUG:TX 4 find_last_bit=2 __fls=2
[   64.680555] DEBUG:RX 2 find_next_bit=1 __ffs=1
[   64.680898] DEBUG:TX 8 find_next_bit=3 __ffs=3
[   64.680899] DEBUG:TX 8 find_last_bit=3 __fls=3
[   64.682640] DEBUG:RX 1 find_next_bit=0 __ffs=0
[   64.682826] DEBUG:TX 1 find_next_bit=0 __ffs=0
[   64.682827] DEBUG:TX 1 find_last_bit=0 __fls=0
[   64.684251] DEBUG:RX 2 find_next_bit=1 __ffs=1
[   64.684584] DEBUG:TX 2 find_next_bit=1 __ffs=1
[   64.684585] DEBUG:TX 2 find_last_bit=1 __fls=1
```

File: sound/soc/codecs/rt1011.c (hunk 6 above)
Comment (ranj063) [reply sound/soc/codecs/rt1011.c]: @lyakh @fredoh9 if we change to use ffs() , then do we need to make rx_mask_l/tx_mask_l unsigned long's? We could just use rx_mask and tx_mask directly no?

Comment (lyakh) [reply sound/soc/codecs/rt1011.c]: @ranj063 @fredoh9 no, we don't need those, that was the idea, sorry for not explaining

**Comment by fredoh9** [comment]
@plbossart @ranj063 
From @lyakh suggestion, replaced with __ffs()/__fls(). Much simpler.

File: sound/soc/codecs/rt1011.c (hunk 7 above)
Comment (fredoh9) [reply sound/soc/codecs/rt1011.c]: if the mask is 0, then return value is not 0? Should we check mask against 0 first? Found a comment in builtin-__ffs.h
`* Undefined if no bit exists, so code should check against 0 first.`

Comment (fredoh9) [reply sound/soc/codecs/rt1011.c]: Comparing __ffs/__fls and ffs/fls. __ffs/__fls has -1(to make 0 index?) that why input param shouldn't be 0. Not sure we can safely assume `mask` cannot be 0. Diving into the code more.

```
[  106.575962] DEBUG:RX 1 __ffs=0 ffs=1
[  106.576150] DEBUG:TX1 4 __ffs=2 ffs=3
[  106.576151] DEBUG:TX2 4 __fls=2 fls=3
[  106.577847] DEBUG:RX 2 __ffs=1 fls=2
[  106.578192] DEBUG:TX1 8 __ffs=3 ffs=4
[  106.578193] DEBUG:TX2 8 __fls=3 fls=4
[  106.580044] DEBUG:RX 1 __ffs=0 ffs=1
[  106.580224] DEBUG:TX1 1 __ffs=0 ffs=1
[  106.580225] DEBUG:TX2 1 __fls=0 fls=1
[  106.581775] DEBUG:RX 2 __ffs=1 ffs=2
[  106.582109] DEBUG:TX1 2 __ffs=1 ffs=2
[  106.582111] DEBUG:TX2 2 __fls=1 fls=2
```

Comment (fredoh9) [reply sound/soc/codecs/rt1011.c]: I make sure input is not 0

File: sound/soc/codecs/rt1011.c (hunk 6 above)
Comment (fredoh9) [reply sound/soc/codecs/rt1011.c]: removed temporary variables, we don't need them.

**Comment by fredoh9** [comment]
I worked on this in this morning. Just found I didn't push my last change. Sorry just pushed now.

File: sound/soc/codecs/rt1011.c [hunk 8]
```diff
@@ -1892,11 +1892,17 @@ static int rt1011_set_tdm_slot(struct snd_soc_dai *dai,
...
+		dev_dbg(component->dev, "too many tx slots or zero slot\n");
+		goto _set_tdm_err_;
+	}
+
+	first_bit = __ffs(tx_mask);
+	last_bit = __fls(tx_mask);
+	if ((last_bit - first_bit) > 1) {
+		ret = -EINVAL;
+		dev_dbg(component->dev, "tx slot location error\n");
 		goto _set_tdm_err_;
```
Comment (ranj063) [inline sound/soc/codecs/rt1011.c:1906]: @fredoh9 I know it was a debug log even before but if we're returning na negative value, shouldn't it be dev_err? Same comment above too

Comment (fredoh9) [reply sound/soc/codecs/rt1011.c:1906]: Yes, makes sense. Testing again with dev_err().

File: sound/soc/codecs/rt1011.c [hunk 9]
```diff
@@ -1892,11 +1892,17 @@ static int rt1011_set_tdm_slot(struct snd_soc_dai *dai,
...
+	if (tx_slotnum > 2 || tx_slotnum == 0) {
 		ret = -EINVAL;
-		dev_dbg(component->dev, "too many tx slots or tx slot location error\n");
+		dev_err(component->dev, "too many tx slots or zero slot\n");
+		goto _set_tdm_err_;
+	}
+
+	first_bit = __ffs(tx_mask);
+	last_bit = __fls(tx_mask);
+	if ((last_bit - first_bit) > 1) {
```
Comment (lyakh) [inline sound/soc/codecs/rt1011.c]: I see, that it was there before too, but since you're at it, please also remove the superfluous parentheses around subtraction.

File: sound/soc/codecs/rt1011.c [hunk 10]
```diff
@@ -1892,11 +1892,17 @@ static int rt1011_set_tdm_slot(struct snd_soc_dai *dai,
 
 	/* Tx slot configuration */
 	tx_slotnum = hweight_long(tx_mask);
-	first_bit = find_next_bit((unsigned long *)&tx_mask, 32, 0);
-	last_bit = find_last_bit((unsigned long *)&tx_mask, 32);
-	if (tx_slotnum > 2 || (last_bit-first_bit) > 1) {
+	if (tx_slotnum > 2 || tx_slotnum == 0) {
```
Comment (lyakh) [inline sound/soc/codecs/rt1011.c]: checkpatch *might* propose `!tx_slotnum` instead of `== 0` but personally I don't care that much about these

File: sound/soc/codecs/rt1011.c (hunk 9 above)
Comment (fredoh9) [reply sound/soc/codecs/rt1011.c]: yeah. I tried to keep previous code as much as possible and also wanted to keep parentheses 
 when operator precedence is sometimes confusing. But I think this is good point also, hence removed superfluous parentheses.

**Comment by kv2019i** [comment]
Thanks @fredoh9 !
